### PR TITLE
WIP 5539: Modularize burn searches

### PR DESCRIPTION
**Pull request by rseanhall**
For https://github.com/wixtoolset/issues/issues/5539. I don't see this getting implemented until after everything's been moved to its own repo.

### CODE REVIEW

File: src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        </xs:annotation>
+        <xs:complexType>
+            <xs:attribute name="Id" type="xs:string" />
+            <xs:attribute name="Variable" type="xs:string" />
+            <xs:attribute name="Condition" type="xs:string" />
+            <xs:attribute name="ExtensionId" type="xs:string" />
+        </xs:complexType>
+    </xs:element>
+
+4) Move the `Util` extension's searches out of the Burn engine and into this new `ExtensionSearch` framework.
```
Comment (barnson) [inline src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md]: I'd start by moving them to dutil; that would simplify creating MSI custom actions out of the searches, something I know has been requested but cannot find an issue for atm.

File: src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+5) Move the `WixFileSearch`, `WixRegistrySearch`, `WixComponentSearch`, and `WixProductSearch` tables out of Wix's tables.xml and into the `UtilExtension`'s tables.xml.
+
+6) In `CreateBurnManifestCommand`, it needs to somehow involve the extension in creating the xml.
+Ideally the extension would somehow register a table as a Burn search table, and be able to provide its own subclass of `Row` for each of the rows.
+This class would inherit from a new interface (`ISearchRow`?) which would have a method that returned an instance of `WixSearchInfo`.
+
+However, this ability for an extension to provide its own subclass of `Row` does not exist yet (this was first brought up [here](https://github.com/wixtoolset/wix4/pull/85)). So the alternative would be to create a new Wix extension interface like `IBurnBinderExtension` and get `CreateBurnManifestCommand` to call each of those extensions and generate the xml that way.
+
```
Comment (barnson) [inline src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md]: `IBurnBackendExtension` exists but there's no manifest extensibility yet.

File: src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    </BurnExtension>
+    <ExtensionSearch Id="myid" Variable="InstallFolder" Condition="PreviousInstallFolder" ExtensionId="Util">
+        <DirectorySearch Path="[PreviousInstallFolder]" Type="path" />
+    </ExtensionSearch>
+
+## Considerations
+
+Once this functionality is in place, I anticipate two feature requests:
+
+1) Add the option to run the search in the elevated process in order to do IIS searches.
```
Comment (barnson) [inline src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md]: Plan still happens in the unelevated process.

File: src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+2) Add the ability to add custom package types and execute them in the Burn extension.
```
Comment (barnson) [inline src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md]: Ooh. Oh. No. No?

File: src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md (hunk 1 above)
Comment (rseanhall) [reply src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md]: Sounds like a good idea. I was more interested in outlining the Burn specific code than where the actual code will be moved.

File: src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md (hunk 2 above)
Comment (rseanhall) [reply src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md]: Oh, didn't realize that was created. I guess I need to figure out what Rob's been doing in those micro repos.

File: src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md (hunk 3 above)
Comment (rseanhall) [reply src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md]: I mean, that's a true statement. But searches happen during Detect so... not sure where you were going with that?

File: src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md (hunk 4 above)
Comment (rseanhall) [reply src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md]: I was just trying to anticipate the request. I wasn't trying to making a comment on whether or not it should be done. 😃

File: src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md (hunk 2 above)
Comment (barnson) [reply src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md]: You'll recognize chunks but lots of new stuff...

File: src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md (hunk 3 above)
Comment (barnson) [reply src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md]: Today, Detect and Plan happen before the elevated process exists. So running Detect from the elevated process is another big change.

File: src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md (hunk 4 above)
Comment (barnson) [reply src/Web/Static/documents/development/wips/5539-modularize-burn-searches.html.md]: Yeah, gotta think about that. It's almost appealing. Almost.

**Pull request merged**